### DISCUSSION TITLE
How to install and get going the project?

### DISCUSSION THREAD

**Discussion by guilleab5** [GetDKAN/dkan#4089] [Q&A]
Hello, How can I get a clean install of the project? Just git cloning it? Or do I have to do a composer create-project or something like that? Sorry, I'm new to dkan. If you have a guide or something that will be much help. Thank you in advance.

P.D.: I tried to follow the documentation but there's only a suggestion to use the getdkan/recommended-project, I don't know if I should just use that?

**Comment by jastraat** [comment]
Hi guilleab5, the getdkan/recommended-project is essentially a composer file that includes all of DKAN's dependencies, DKAN, and a patch for Drupal core. The easiest way to get an initial installation working locally is with DDEV + the DDEV-DKAN add-on: https://getdkan.github.io/ddev-dkan/getting-started.html

**Comment by janette** [comment]
See also https://dkan.readthedocs.io/en/latest/installation/index.html#starting-a-new-dkan-project

**Comment by WaudoDavid** [comment]
How can one use dkan in production without the DDEV as it seems to be used locally only?

**Reply by dmundra** [reply]
I believe you can install DKAN via composer. Repeat Janette's link https://dkan.readthedocs.io/en/latest/installation/index.html#adding-dkan-into-an-existing-project

**Reply by WaudoDavid** [reply]
Thank you for your response. I have tried using the command composer require 'getdkan/dkan' on my already existing drupal site but I keep getting dependency issues in the composer.json that are mismatching furthermore I understand that this dependency **npm-asset/select2:^4.0** doesn't exist according to this issue raised https://github.com/GetDKAN/dkan/issues/3740

**Reply by dmundra** [reply]
Ah got it. Can you share here or raise an issue sharing the dependency issues you are seeing? Can you share your composer.json?

**Reply by WaudoDavid** [reply]
This is my composer.json for my drupal site and am using this image of drupal **drupal:php8.3-apache-bullseye**

{
    "name": "drupal/recommended-project",
    "description": "Project template for Drupal projects with a relocated document root",
    "type": "project",
    "license": "GPL-2.0-or-later",
    "homepage": "https://www.drupal.org/project/drupal",
    "support": {
        "docs": "https://www.drupal.org/docs/user_guide/en/index.html",
        "chat": "https://www.drupal.org/node/314178"
    },
    "repositories": [
        {
            "type": "composer",
            "url": "https://packages.drupal.org/8"
        }
    ],
    "require": {
        "composer/installers": "^2.0",
        "drupal/core-composer-scaffold": "^11.0",
        "drupal/core-project-message": "^11.0",
        "drupal/core-recommended": "^11.0"
    },
    "conflict": {
        "drupal/drupal": "*"
    },
    "minimum-stability": "dev",
    "prefer-stable": true,
    "config": {
        "allow-plugins": {
            "composer/installers": true,
            "drupal/core-composer-scaffold": true,
            "drupal/core-project-message": true,
            "phpstan/extension-installer": true,
            "dealerdirect/phpcodesniffer-composer-installer": true,
            "php-http/discovery": true
        },
        "sort-packages": true
    },
    "extra": {
        "drupal-scaffold": {
            "locations": {
                "web-root": "web/"
            }
        },
        "installer-paths": {
            "web/core": [
                "type:drupal-core"
            ],
            "web/libraries/{$name}": [
                "type:drupal-library"
            ],
            "web/modules/contrib/{$name}": [
                "type:drupal-module"
            ],
            "web/profiles/contrib/{$name}": [
                "type:drupal-profile"
            ],
            "web/themes/contrib/{$name}": [
                "type:drupal-theme"
            ],
            "drush/Commands/contrib/{$name}": [
                "type:drupal-drush"
            ],
            "web/modules/custom/{$name}": [
                "type:drupal-custom-module"
            ],
            "web/profiles/custom/{$name}": [
                "type:drupal-custom-profile"
            ],
            "web/themes/custom/{$name}": [
                "type:drupal-custom-theme"
            ]
        },
        "drupal-core-project-message": {
            "include-keys": [
                "homepage",
                "support"
            ],
            "post-create-project-cmd-message": [
                "<bg=blue;fg=white>                                                         </>",
                "<bg=blue;fg=white>  Congratulations, you’ve installed the Drupal codebase  </>",
                "<bg=blue;fg=white>  from the drupal/recommended-project template!          </>",
                "<bg=blue;fg=white>                                                         </>",
                "",
                "<bg=yellow;fg=black>Next steps</>:",
                "  * Install the site: https://www.drupal.org/docs/installing-drupal",
                "  * Read the user guide: https://www.drupal.org/docs/user_guide/en/index.html",
                "  * Get support: https://www.drupal.org/support",
                "  * Get involved with the Drupal community:",
                "      https://www.drupal.org/getting-involved",
                "  * Remove the plugin that prints this message:",
                "      composer remove drupal/core-project-message"
            ]
        }
    }
}

**Reply by WaudoDavid** [reply]
I get this error 

Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Root composer.json requires getdkan/dkan 2.x-dev -> satisfiable by getdkan/dkan[2.x-dev].
    - getdkan/dkan 2.x-dev requires npm-asset/select2 ^4.0 -> could not be found in any version, there may be a typo in the package name.

**Reply by dmundra** [reply]
Ah I see the issue, we have not updated DKAN to support Drupal 11 just yet, see https://github.com/GetDKAN/dkan/discussions/4294 for that discussion. It will be our focus sometime in the future. Right now we do support up to 10.3 and PHP 8.3. As Paul mentions pull requests to help us get to Drupal 11 would be appreciated.

**Reply by dmundra** [reply]
I am testing `composer require 'getdkan/dkan:2.x-dev'` and can confirm your error.

**Reply by WaudoDavid** [reply]
what about that dependency **npm-asset/select2 ^4.0** because it would still be an issue when I run  **composer require getdkan/dkan:2.x-dev**

**Reply by WaudoDavid** [reply]
What would the solutione be?

**Reply by dmundra** [reply]
Yes the solution for 'npm-asset/select2 ^4.0' is adding this repository in the 'repositories' section of composer.json (see https://github.com/hiqdev/asset-packagist for details). So yours would look like:

```
"repositories": [
{
"type": "composer",
"url": "https://packages.drupal.org/8"
},
{
"type": "composer",
"url": "https://asset-packagist.org"
}
]
```

**Reply by WaudoDavid** [reply]
Thank you so much @dmundra I was able to succesfully install it. I would love to reach out to you again if I come across any other issue.

**Reply by dmundra** [reply]
Glad to hear it and sounds good.

**Reply by WaudoDavid** [reply]
Hello once again @dmundra. How do I install the decoupled frontend app in my drupal site. I have followed this steps in https://github.com/GetDKAN/data-catalog-app on how to **Manual Set Up in a DKAN site** but I can't seem to see the default https://demo.getdkan.org/ the way it appears here as my homepage

**Reply by dmundra** [reply]
I am new to this portion but I followed the steps and I was able to get the front end loaded on my Drupal site. Key points I ran into:
* Making sure the cloned frontend folder is in the same folder as the Drupal files, in my case it was in the 'web' folder. Yours could be 'docroot' or something else.
* Run `npm run build` in the frontend folder to build the files.
* If you don't change anything in the module or frontend, go to your site /home and the frontend should load.

**Reply by WaudoDavid** [reply]
@dmundra what kind of drupal installation do you have. Would you please share that and also did you install it in a docker container or vm. 

Am using a docker container and my config is as follows:

services:
  drupal:
    container_name: drupal
    image: drupal:10.0-php8.2-apache
    ports:
      - "8100:80"    
    volumes:
      - drupal_modules:/var/www/html/modules
      - drupal_profiles:/var/www/html/profiles
      - drupal_themes:/var/www/html/themes
      - drupal_sites:/var/www/html/sites
      - ./frontend:/opt/drupal/web/frontend  # Mount the frontend directory into /opt/drupal/frontend   
    restart: always
    depends_on:
      pilgrim-dkan-db2:
        condition: service_healthy  # Wait for MySQL to be healthy
    networks:
      - pilgrim_dkan_network

  pilgrim-dkan-db2:
    image: mysql:8.0.13
    container_name: pilgrim-dkan-db2
    environment:
      MYSQL_ROOT_PASSWORD: password
      MYSQL_DATABASE: dkan
      MYSQL_USER: user
      MYSQL_PASSWORD: password
    volumes:
      - mysql_data:/var/lib/mysql
    restart: always
    healthcheck:
      test: ["CMD", "mysqladmin", "ping", "-h", "localhost"]
      interval: 30s
      timeout: 10s
      retries: 5
    networks:
      - pilgrim_dkan_network

networks:
  pilgrim_dkan_network:
    driver: bridge
volumes:
  drupal_modules:
  drupal_profiles:
  drupal_themes:
  drupal_sites:
  mysql_data:

**Reply by WaudoDavid** [reply]
@dmundra Also in your installation are you able to log in to drupal by going to your /user/login since you have installed the frontend?

**Reply by dmundra** [reply]
@WaudoDavid I am using a docker container that is very similar to your setup.

> Also in your installation are you able to log in to drupal by going to your /user/login since you have installed the frontend?

I was already logged in so I need to try what that looks like.

**Reply by dmundra** [reply]
For my setup when I go to /home I get the DKAN frontend. When I got to /user/login I get the Drupal login just fine.

**Reply by WaudoDavid** [reply]
@dmundra would you please share your setup step by step so that I can have a look. That would be much appreciated.

**Reply by dmundra** [reply]
Sure. Looks like the Drupal docker image is to limiting to work with (I tried to set it up using your docker-compose file). Here is my docker setup:

docker-compose.yml, note that I have the full Drupal site installed in the same folder as this docker-compose.yml file and mount db folder as volume for mysql.

```
services:
  web:
    build:
      context: ./
      dockerfile: docker/Dockerfile
    volumes:
      - .:/var/www
    ports:
      - 8011:80

  db:
    image: mysql:5.7
    environment:
      MYSQL_DATABASE: drupal
      MYSQL_USER: drupal
      MYSQL_PASSWORD: drupal
      MYSQL_ROOT_PASSWORD: drupal
    ports:
      - 3311:3306
    command: mysqld --max_allowed_packet=32M --lower_case_table_names=1
    volumes:
      - ./db:/var/lib/mysql
```

Dockerfile located in a folder called 'docker':

```
FROM php:8.3-apache
ENV TZ=America/Los_Angeles
RUN ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone
RUN a2enmod rewrite
RUN a2enmod expires
COPY docker/custom-config.ini /usr/local/etc/php/conf.d
RUN service apache2 restart
RUN apt-get update && apt-get install -y \
        libfreetype6-dev \
        libjpeg62-turbo-dev \
        libpng-dev \
        libwebp-dev \
        libtidy-dev \
        libzip-dev \
        zip \
        libonig-dev \
        libicu-dev \
    && docker-php-ext-install -j$(nproc) iconv pdo_mysql tidy intl \
    && docker-php-ext-configure gd --with-freetype --with-jpeg --with-webp \
    && docker-php-ext-install -j$(nproc) gd mysqli \
    && docker-php-ext-install zip

# Used by drush to bootstrap drupal
RUN apt-get update && apt-get install -y default-mysql-client git unzip

# For Drupal 8
RUN docker-php-ext-install opcache
COPY docker/opcache-recommended.ini /usr/local/etc/php/conf.d

ENV APACHE_DOCUMENT_ROOT /var/www/web

RUN sed -ri -e 's!/var/www/html!${APACHE_DOCUMENT_ROOT}!g' /etc/apache2/sites-available/*.conf
RUN sed -ri -e 's!/var/www/!${APACHE_DOCUMENT_ROOT}!g' /etc/apache2/apache2.conf /etc/apache2/conf-available/*.conf
```

custom-config.ini

```
; Error Handling
log_errors = On
display_errors = On
error_reporting = E_ALL & ~E_DEPRECATED & ~E_STRICT

; Memory Handling
max_execution_time = 120
max_input_time = 120
memory_limit = 128M
upload_max_filesize = 64M
post_max_size = 64M

; Timezone
date.timezone = America/Los_Angeles
```

opcache-recommended.ini

```
; Opcache Recommendation for Drupal 8
opcache.memory_consumption = 128
opcache.interned_strings_buffer = 8
opcache.max_accelerated_files = 4000
opcache.revalidate_freq = 60
opcache.fast_shutdown = 1
```

With the above I can run composer and drush commands either in the docker container or outside to install/update dependencies as needed.

I think you still have to take the above and make it work for your setup.

**Reply by WaudoDavid** [reply]
Have you set up DKAN on this docker container? Please if you have share your code that you have added to get it runnin even with the frontend?

**Reply by dmundra** [reply]
Yes and I did follow the same steps as in the documentation to update composer.json with the additions to repositories section and running `composer require 'getdkan/dkan:2.x-dev'` and following the frontend steps in the documentation for that. Nothing in the code to show.

```
    "repositories": [
        {
            "type": "composer",
            "url": "https://packages.drupal.org/8"
        },
        {
            "type": "composer",
            "url": "https://asset-packagist.org"
        }
    ],
```

**Reply by WaudoDavid** [reply]
**Manual Set Up in a DKAN site**
If you have a backend already running and just need the frontend:

1.) Clone this repository in your docroot git clone https://github.com/getdkan/data-catalog-app frontend. The DKAN DDEV add-on is structured to run commands in a folder named frontend so frontend repos could be swapped if needed.
2.) Install the dependencies with [npm](https://www.npmjs.com/):
    i.) cd frontend
    ii.) npm install or yarn install
3.) Run the server: npm start or yarn start
    i.)Your site is now running at http://localhost:3000
4.) Build the public files npm run build

5.) To complete the setup like the auto setup, setup the DKAN JS Frontend module that comes with the DKAN core installation. Steps can be found here: [DKAN JS Frontend Module](https://github.com/GetDKAN/dkan/tree/2.x/modules/dkan_js_frontend).

*******************************************************************************************************************

From the instructions above this is what I understand and I have done the following:

1.) I have installed the DKAN module as intended and enabled the DKAN JS Frontend module that comes with the DKAN core installation.

2.) I have cloned the repo in my folder for web where there are other folders like the _core, sites, themes_ alongside it and run cd frontend where I have installed the dependencies with **npm install.**

3.) Finally I have built the public files with **npm run buid**


Am using this version of dkan and Web Server Apache/2.4.62 (Debian) and Drupal Version 10.2.2
 {
            "name": "getdkan/dkan",
            "version": "2.17.0",
            "source": {
                "type": "git",
                "url": "https://github.com/GetDKAN/dkan.git",
                "reference": "a8c2f44c2d17cc21726d056d3faec833cfd72dfd"
 },



Is there something am missing? Please correct me if I'm wrong. Thank you so much for your time.

**Reply by dmundra** [reply]
I think that is all fine. Note that I am using the latest version of dkan. Are you getting an error when you try to load your site at the URL /home?

**Reply by WaudoDavid** [reply]
The DKAN frontend is being displayed but I can't go to the admin dashboard by adding the path /url/login.

**Reply by dmundra** [reply]
Interesting, what happens when you do that? Also I presume you are trying the URL `/user/login`? If you try a new browser do you have the same issue?
